File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/card/infrastructure/AccountAdapterImpl.kt
package com.example.accountservice.card.infrastructure

import com.example.accountservice.account.application.query.AccountQuery
import com.example.accountservice.account.domain.AccountFindQuery
import com.example.accountservice.account.domain.AccountStatus
import com.example.accountservice.card.application.adapter.AccountAdapter
import com.example.accountservice.card.application.adapter.AccountView
import org.springframework.stereotype.Component

/**
 * Implementation of [AccountAdapter] (ACL). Injects and calls the read port ([AccountQuery]) exposed
 * by the Account BC, and translates the Account BC's model
 * ([com.example.accountservice.account.domain.Account]/[AccountStatus]) into the minimal form the Card
 * BC uses ([AccountView]). It does not reference Account's write Repository or domain methods.
 *
 * Account's "account not found" signal is `AccountQuery.findAccounts` returning an empty list (single
 * lookups are also handled via `take = 1` + `firstOrNull()` — repository-pattern.md), and this is
 * propagated as-is to `null`, which the Card domain understands — Account's exception types (such as
 * AccountNotFoundException) never leak into the Card layer.
 */
@Component
class AccountAdapterImpl(
    private val accountQuery: AccountQuery,
) : AccountAdapter {
    override fun findAccount(
        accountId: String,
        ownerId: String,
    ): AccountView? {
        val (accounts, _) =
            accountQuery.findAccounts(AccountFindQuery(page = 0, take = 1, accountId = accountId, ownerId = ownerId))
        return accounts.firstOrNull()?.let { account ->
            AccountView(accountId = account.accountId, active = account.status == AccountStatus.ACTIVE, email = account.email)
        }
    }
}

File: implementations/fastapi/examples/src/payment/infrastructure/account_adapter_impl.py
from ...account.domain.account_status import AccountStatus
from ...account.domain.repository import AccountQuery
from ..application.adapter.account_adapter import AccountAdapter, AccountView


class AccountAdapterImpl(AccountAdapter):
    """The implementation of AccountAdapter (ACL). It's injected with and calls the read
    interface (AccountQuery) the Account BC exposes, and translates Account BC's
    model/status into the minimal shape the Payment BC uses (AccountView). It never
    references Account's Repository implementation or domain objects directly.
    """

    def __init__(self, account_query: AccountQuery) -> None:
        self._account_query = account_query

    async def find_account(self, account_id: str, owner_id: str) -> AccountView | None:
        accounts, _ = await self._account_query.find_accounts(page=0, take=1, account_id=account_id, owner_id=owner_id)
        account = accounts[0] if accounts else None
        if account is None:
            return None
        return AccountView(
            account_id=account.account_id,
            active=account.status == AccountStatus.ACTIVE,
            balance_amount=account.balance.amount,
            currency=account.balance.currency,
        )
